docs/architecture.md

```diff
@@ -0,0 +1,292 @@
+management, and deletion of Spark Executors tied to a particular Driver program.
+
+### Integrating with `spark-submit`
+
+While Spark seems to take some care around providing a defined interface for
+individual Cluster Managers, there's still a fair bit of cluster-specific code
+within `SparkSubmit`. Absent upstream integration, `armada-spark` maintains and
+provides it's own `ArmadaSparkSubmit` implementation which serves as a drop-in
+replacement.
+
+Currently the following versions of Spark are supported:
+- 3.3
+- 3.5
+- 4.1
+
+### Using Armada as a "cluster"
+
+While Armada can be treated as a cluster for the purposes of Spark, some care
+and consideration must be taken in order to meet most, if not all, of Spark's
+expectations.
+
+### Driver and Executor (and Cluster Manager) Network Communications
+
+One major requirement that **must** be met is network communication between
+Drivers and Executors. This allows Drivers to submit tasks to the Executors
+and collate their results. `armada-spark` achieves this through the use of
+[kubernetes labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/).
+Used correctly, these labels induce the Armada scheduler to schedule Drivers
+and their accompanying Executors to the same kubernetes cluster managed by
+Armada. Kubernetes cluster networking topology guarantees that nodes within the
+cluster will be able to communicate. Scheduling to the same cluster also has
+the benefit of increasing the probability of compute resources being
+"clustered" closely together. Perhaps servers on the same rack, or even a single
+server.
+
+If you refer back to the ["Cluster Manager Overview"](https://spark.apache.org/docs/latest/cluster-overview.html),
+the first image you'll find is a diagram showing a complete, directed graph between the:
+
+- Driver Program
+- Worker Node / Executors
+- Cluster Manager
+
+and while the diagram appears to separate the "Cluster Manager" out as a separate
+program or node, in reality it tends to "live" in the same process as the
+Driver. So by ensuring connectivity between the Driver and Executors, we also
+get connectivity with the "Cluster Manager".
+
+### Driver and Executors as Armada Jobs
+
+Armada, being a multi-cluster Kubernetes manager, takes on the role of
+scheduling "jobs" to the clusters it oversees. For now, it makes the most sense
+to separate each Driver and Executor into their own Armada job. There are 
+several considerations that must be observed when submitting these jobs.
+
+#### Armada Gang Scheduling
+
+`armada-spark` uses gang scheduling annotations to schedule the driver and 
+its executors together. By using the gang scheduling “Node Uniformity Label” 
+we can guarantee that a driver and its executors are scheduled to the same
+cluster if Kubernetes node labels are applied correctly.
+
+#### Kubernetes Node Labels
+
+Any Kubernetes worker nodes should have a common label, such as "armada-spark", 
```

Comment:
Is this still needed with Gang scheduling?

Reply:
Yes, gang scheduling relies on Kubernetes labels to function.

Reply:
Maybe I'm misremembering, but I thought the reason we added gang scheduling was because @dave-gantenbein didn't want to force the armada ops team to have to add these labels?

Reply:
If these labels are required on each worker node then why do we need gang-scheduling?

Reply:
Isn't it because the Spark user will specify some value of `spark.executor.instances` and we have to adhere to that, otherwise job performance (e.g. results-shuffling operations) would vary wildly unless we can gang-schedule pods all at once?

Reply:
I mean, I questioned the need for gang scheduling before implementing it. But gang scheduling can be slightly more convenient for us *if* it ends up supporting dynamic addition/removal of jobs. 

Gang scheduling already depends on Kubernetes labels, at least if you use the `gangNodeUniformityLabel` annotation, which we do and we should because we want to guarantee some type of computational/network locality for performance's sake (not to mention single-cluster).

Reply:
My point is that as long as we have gang scheduling, the worker nodes don't need a common label, as far as I know.  In @dejanzele 's PR:   one or the other is required, not both: https://github.com/armadaproject/armada-spark/blob/18b4574d8b5bf3b2d846f9ceeddd629f969641ca/src/main/scala/org/apache/spark/deploy/armada/submit/ArmadaClientApplication.scala#L171

Reply:
am i misunderstanding something?

Reply:
The worker nodes absolutely need a label whether we use gang scheduling or not. We still want to guarantee single-cluster at least. It's possible I'm misunderstanding something but I don't think gang scheduling guarantees single cluster on its own.

Reply:
Ok thanks for clarifying, sorry for the misunderstanding.